File: Typescript/session1.td/Zero.ts
import { Nat } from "./Nat";
import { Succ } from "./Succ"
import { NatParInt } from "./NatParInt"
import { FabriqueNat } from "./FabriqueNat"

export class Zero implements Nat{
    public static FAB:FabriqueNat<Nat>=new NatParInt(0);
    estNul(): boolean {
        return true;
    }    
    predecesseur(): Nat {
        throw new Error("Pas de prédecesseur entier naturel à 0.");
    }
    chiffre(value: number): number {
        return 0;
    }
    taille(): number {
        return 1;
    }
    val(): number {
        return 0;
    }
    creerNatAvecValeur(value: number): Nat {
        if(value==0){
            return new Zero();
        } else {
            return NatParInt.FAB.creerNatAvecValeur(value);
        }
    }
    creerZero(): Nat {
        return this;
    }
    creerSuccesseur(nat: Nat): Nat {
        return new Succ(nat);
    }
    creerNatAvecRepresentation(chaine: string): Nat {
        if(Number(chaine)<0){
            throw new Error("Le nombre n'est pas un entier naturel (Zero creeRep)");
        } else {
            return this.creerNatAvecValeur(Number(chaine));
        }
    }
    somme(x: Nat): Nat {
        return this.creerNatAvecValeur(x.val());
    }
    zero(): Nat {
        return this.creerZero();
    }
    produit(x: Nat): Nat {
        return this.creerZero();
    }
    un(): Nat {
        return this.creerNatAvecValeur(1);
    }
    modulo(x: Nat): Nat {
        return this.creerZero();
    }
    div(x: Nat): Nat {
        return this.creerZero();
    }

    equals(x: Nat): boolean {
        return x.val()==0;
    }
    toString():string{
        return String(this.val());
    }


}

File: Java/src/session1/td/Zero.java
package session1.td;

public class Zero implements Nat{
	static final FabriqueNat<Nat> FAB = new Zero();
	
	public Zero() {}

	@Override
	public Nat creerNatAvecValeur(int x) {
		if(x==0) {
			return this.creerZero();
		}else {
			//return this.creerSuccesseur(this.creerNatAvecValeur(x - 1));
			return NatParInt.FAB.creerNatAvecValeur(x);
		}
	}

	@Override
	public boolean estNul() {
		return true;
	}

	@Override
	public Nat predecesseur() {
		
			throw new java.lang.UnsupportedOperationException();
	}
	
	@Override
	public String toString(){
		return String.valueOf(this.val());
	}
	@Override
	public boolean equals(Object o) {
		//System.out.println("o.val()"+ o.val());
		//System.out.println("this.val()"+this.val());
		return ((Nat)o).val()==this.val();
	}

	@Override
	public Zero creerZero() {
		return this;
	}

	@Override
	public int val() {
		return 0;
	}

	@Override
	public int taille() {
		return 1;
	}

	@Override
	public int chiffre(int i) {
		return 0;
	}

	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return new SuccRec(predecesseur);
	}

	@Override
	public Nat somme(Nat n) {
		int result = this.val() + n.val();
		System.out.println("this Val : " + String.valueOf(this.val()));
		System.out.println("other val : "+ String.valueOf(n.val()));
		return this.creerNatAvecValeur(result);
	}

	@Override
	public Nat zero() {
		return this.creerZero();
	}

	@Override
	public Nat produit(Nat n) {
		return this.creerZero();
	}

	@Override
	public Nat un() {
		return this.creerNatAvecValeur(1);
	}

	@Override
	public Nat modulo(Nat x) {
		return this.creerZero();
	}

	@Override
	public Nat div(Nat x) {
		return this.creerNatAvecValeur(this.val()/x.val());
	}

	
}
